### DISCUSSION TITLE
gRPC flow

### DISCUSSION THREAD

**Discussion by ProjectInitiative** [seaweedfs/seaweedfs#3397] [Q&A] [answered]
Looking up directory and file meta data is pretty straight forward with the proto definitions `LookupDirectoryEntry` and `ListEntries`. 

What is the proper way of requesting file content from the filer gRPC definitions? 

Tired pulling the `content` message from `Entry` however content seems to be empty with every request. 

My guess would be something along the lines of pulling all chunks from metadata and somehow requesting chunks directly from volume servers.

What is the current flow for how `weed mount` pulls data from the filer servers? 

Thanks

**Comment by chrislusf** [comment, marked as answer]
This is the simplest case:
1. Read the entry metadata
2. Fetch chunks via http from volume servers.

There are cases where some chunks may overlap each other.

For files larger than 2GB, some chunks are "meta chunk", which contain a list of chunks.

**Reply by ProjectInitiative** [reply]
Thanks, I will take a look. I am looking to use gRPC as it seems to be faster than HTTP

**Reply by ProjectInitiative** [reply]
@chrislusf are there only HTTP API hooks for reading chunks? Was looking through the protobuf files and mount code to understand the flow a little more.

From what I can see, it looks like `weed mount` communicates with the filer with gRPC, and then requests chunks with HTTP. Is this correct? 

Thanks

**Reply by chrislusf** [reply]
Yes
